### PR TITLE
Allow multiple uploads in a single job

### PR DESCRIPTION
**Pull request by aeisenberg**
They must all have a unique category.

### Merge / deployment checklist

- [x] Confirm this change is backwards compatible with existing workflows.
- [x] Confirm the [readme](https://github.com/github/codeql-action/blob/main/README.md) has been updated if necessary.
- [x] Confirm the [changelog](https://github.com/github/codeql-action/blob/main/CHANGELOG.md) has been updated if necessary.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -2,7 +2,7 @@
 
 ## [UNRELEASED]
 
-No user facing changes.
+- Allow multiple uploads in a single job as long as they have different categories. [#801](https://github.com/github/codeql-action/pull/ç)
```
Comment (adityasharad) [inline CHANGELOG.md:5]: Not quite the pull number you were going for, I think :)
```suggestion
- The `upload-sarif` action now allows multiple uploads in a single job, as long as they have different categories. [#801](https://github.com/github/codeql-action/pull/801)
```

File: src/upload-lib.ts [hunk 2]
```diff
@@ -409,3 +400,19 @@ async function uploadFiles(
...
+
+export function validateUniqueCategory(category: string | undefined) {
+  if (util.isActions()) {
+    // This check only works on actions as env vars don't persist between calls to the runner
+    const sentinelEnvVar = `CODEQL_UPLOAD_SARIF + ${
+      category ? `_${category}` : ""
+    }`;
+    if (process.env[sentinelEnvVar]) {
+      throw new Error(
+        "Aborting upload: only one run of the codeql/analyze or codeql/upload-sarif actions is allowed per job per category. " +
```
Comment (adityasharad) [inline src/upload-lib.ts:412]: Perhaps include the category in the error message, if one is set, or mention the default is being used if none is set.

File: src/upload-lib.ts [hunk 3]
```diff
@@ -409,3 +400,19 @@ async function uploadFiles(
     num_results_in_sarif: numResultInSarif,
   };
 }
+
+export function validateUniqueCategory(category: string | undefined) {
+  if (util.isActions()) {
+    // This check only works on actions as env vars don't persist between calls to the runner
+    const sentinelEnvVar = `CODEQL_UPLOAD_SARIF + ${
+      category ? `_${category}` : ""
```
Comment (adityasharad) [inline src/upload-lib.ts:408]: Is an arbitrary category string valid as an environment variable suffix?

File: src/upload-lib.test.ts [hunk 4]
```diff
@@ -175,3 +175,14 @@ test("populateRunAutomationDetails", (t) => {
   );
   t.deepEqual(modifiedSarif, expectedSarif);
 });
+
+test("validateUniqueCategory", (t) => {
```
Comment (adityasharad) [inline src/upload-lib.test.ts:179]: Can you also test two _different_ categories in these cases, and that it doesn't throw?

**Review by adityasharad — COMMENTED** [review]
Thanks. I didn't know we had this mechanism and it's pretty neat to avoid user errors. Some suggestions for making that experience even nicer.

File: src/upload-lib.ts (hunk 3 above)
Comment (aeisenberg) [reply src/upload-lib.ts:408]: Oh...that's a good point. We'll need to escape this.

File: src/upload-lib.test.ts (hunk 4 above)
Comment (aeisenberg) [reply src/upload-lib.test.ts:179]: This is already being tested since each time you call `validateUniqueCategory`, an env var is created for that call.

**Comment by aeisenberg** [comment]
Failing Debug artifact upload check is unrelated.

**Review by cklin — APPROVED** [review]

**Pull request merged**